parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java

```diff
@@ -72,29 +74,31 @@
     this.options = options;
     this.reader = newReader(file, options);
     MessageType fileSchema = reader.getFileMetaData().getSchema();
-
     boolean hasIds = ParquetSchemaUtil.hasIds(fileSchema);
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
 
     this.projection = hasIds ?
-        ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+            ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
+            ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema);
+
     this.rowGroups = reader.getRowGroups();
     this.shouldSkip = new boolean[rowGroups.size()];
 
     ParquetMetricsRowGroupFilter statsFilter = null;
     ParquetDictionaryRowGroupFilter dictFilter = null;
     if (filter != null) {
-      statsFilter = new ParquetMetricsRowGroupFilter(expectedSchema, filter, caseSensitive);
-      dictFilter = new ParquetDictionaryRowGroupFilter(expectedSchema, filter, caseSensitive);
+      NameMapping nameMapping = MappingUtil.create(ParquetSchemaUtil.convert(fileSchema));
```

Comment:
I think the `nameMapping` should be taken as a parameter in `Parquet` as part of the builder and passed in from there

Reply:
Also, if the `nameMapping` is not passed in and file schema does not have ids. We can infer the nameMapping, similar to what we've done here.  We had done this for Avro https://github.com/apache/incubator-iceberg/pull/580

Reply:
It makes sense if we build name mapping from the expected schema, while here I need to build it from file schema so that we can use the metrics correctly.